src/Spice86.Core/Emulator/OperatingSystem/Devices/ConsoleDevice.cs

```diff
@@ -108,98 +199,259 @@ public override long Position {
         set => throw new NotSupportedException("Console device does not support setting position.");
     }
 
+    // Thunk writer (inspired by the mouse driver approach).
+    //
+    // What it does:
+    // - We emit a tiny far-callable stub in high memory that waits for a keystroke using INT 16h AH=00h.
+    // - When the BIOS keyboard buffer is empty and a DOS function wants an input char, we patch the topmost
+    //   INT 21h IRET frame on the stack so that the IRET returns to this stub instead of the original site.
+    // - The stub enables interrupts (STI) to ensure IRQ1 can feed the BIOS buffer and then executes INT 16h AH=00h.
+    //   Your existing INT 16h handler’s ASM loop does the actual busy-wait safely with IF=1.
+    // - After INT 16h returns with AX=(scan<<8|ascii), the stub restores preserved registers and returns
+    //   to the original IP:CS by pushing the saved FLAGS, CS, and IP so that a final IRET resumes the program.
+    //
+    // Flags clarification:
+    // - On x86, INT pushes FLAGS, CS, IP in that order; thus at the handler entry, memory at SS:SP is:
+    //     [SP+0]=IP, [SP+2]=CS, [SP+4]=FLAGS (top of stack points to IP).
+    // - We read and save that triplet (IP/CS/FLAGS) into our thunk’s data before hijacking the frame.
+    // - We DO NOT modify the saved FLAGS value; we leave it intact so the final IRET in the thunk restores
+    //   the exact FLAGS state the program would have seen if we hadn’t hijacked the return address. This preserves
+    //   Carry Zero etc. set by the DOS function.
+    // - The stub executes STI early to allow interrupts while waiting. STI sets the runtime IF bit immediately
+    //   (subject to one-instruction shadowing semantics), but this does not change the saved FLAGS we push back later.
+    //   In other words, IF is enabled only during the busy-wait window; the FLAGS restored by the final IRET are
+    //   the original ones (origFLAGS), not affected by the STI we executed in the thunk.
+    //
+    // Registers clarification:
+    // - The thunk saves BX, CX, DX, SI, DI, BP and restores them before returning.
+    // - AX is intentionally NOT saved/restored because AX must carry the keystroke result from INT 16h AH=00h
+    //   back to the program (BIOS semantics: AH=scan code, AL=ASCII or 0).
+    //
+    // Segments clarification:
+    // - The thunk temporarily sets DS=CS (push cs / pop ds) so it can read its local saved variables.
+    //   The code does not restore DS here; see reviewer note below.
+    //
+    // Reviewer note:
+    // - This is a minimal thunk to make the wait happen in guest ASM. If you need DS to be strictly preserved
+    //   across the detour for all callers, add push/pop of DS around the DS=CS change.
+    private void EnsureBlockingThunk() {
+        if (_thunkInitialized) {
```

Comment:
Shouldnt ensuring this is called only once be done at init? Like from the constructor?